src/Lecture2.hs

```diff
@@ -1,170 +1,234 @@
-    , dropSpaces
-
-    , Knight (..)
-    , dragonFight
-
-      -- * Hard
-    , isIncreasing
-    , merge
-    , mergeSort
-
-    , Expr (..)
-    , Variables
-    , EvalError (..)
-    , eval
-    , constantFolding
-    ) where
+  ( -- * Normal
+    lazyProduct,
+    duplicate,
+    removeAt,
+    evenLists,
+    dropSpaces,
+    Knight (..),
+    dragonFight,
+
+    -- * Hard
+    isIncreasing,
+    merge,
+    mergeSort,
+    Expr (..),
+    Variables,
+    EvalError (..),
+    eval,
+    constantFolding,
+  )
+where
 
 -- VVV If you need to import libraries, do it after this line ... VVV
+import Data.Char
 
--- ^^^ and before this line. Otherwise the test suite might fail  ^^^
-
-{- | Implement a function that finds a product of all the numbers in
-the list. But implement a lazier version of this function: if you see
-zero, you can stop calculating product and return 0 immediately.
+-- ^ ^^ and before this line. Otherwise the test suite might fail  ^^^
 
->>> lazyProduct [4, 3, 7]
-84
--}
+-- | Implement a function that finds a product of all the numbers in
+-- the list. But implement a lazier version of this function: if you see
+-- zero, you can stop calculating product and return 0 immediately.
+--
+-- >>> lazyProduct [4, 3, 7]
+-- 84
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
-
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
+lazyProduct [] = 1
+lazyProduct (0 : _) = 0
+lazyProduct (h : t) = h * lazyProduct t
+
+-- lazyProduct =
+--   iter id
+--   where
+--     iter k [] = k 1
+--     iter k (h : t) = iter (k . \x -> h * x) t
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate =
+  let dupleIter [] = []
+      dupleIter (h : t) = h : h : dupleIter t
+   in dupleIter
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Integer -> [a] -> (Maybe a, [a])
+removeAt =
+  removeIter (Nothing, [])
+  where
+    removeIter (x, l) _ [] = (x, reverse l)
+    removeIter (x, l) count (h : t)
+      | count < 0 = (Nothing, h : t)
+      | count == 0 = (Just h, reverse l ++ t)
+      | otherwise = removeIter (x, h : l) (count - 1) t
+
+-- | Write a function that takes a list of lists and returns only
+-- lists of even lengths.
+--
+-- >>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
+-- [[3,1,2,7],[]]
+--
+-- ♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
+--  in this function.
+evenLists :: [[a]] -> [[a]]
+evenLists =
+  let evenfilter l = [i | i <- l, (even . length) i]
+   in evenfilter
```

Comment:
Nice usage of list comprehension!

Alternatively, you can use the `filter` function on the top level to implemet this function concisely
```suggestion
evenLists = filter (even . length)
```